packages/react-storage/src/components/StorageBrowser/Views/LocationsView/Controls/DataTable.tsx

```diff
@@ -0,0 +1,173 @@
+      onClick: () => onTableHeaderClick('scope'),
+      'aria-sort': selection === 'scope' ? direction : 'none',
+    },
+    {
+      children: (
+        <ButtonElement
+          variant="sort"
+          className={TABLE_HEADER_BUTTON_CLASS_NAME}
+        >
+          Type
+          <IconElement
+            variant={
+              selection === 'type' && direction !== 'none'
+                ? `sort-${direction}`
+                : 'sort-indeterminate'
+            }
+          />
+        </ButtonElement>
+      ),
+      key: 'th-type',
+      className: `${TABLE_HEADER_CLASS_NAME} ${TABLE_HEADER_CLASS_NAME}--type`,
+      onClick: () => onTableHeaderClick('type'),
+      'aria-sort': selection === 'type' ? direction : 'none',
+    },
+    {
+      children: (
+        <ButtonElement
+          variant="sort"
+          className={TABLE_HEADER_BUTTON_CLASS_NAME}
+        >
+          Permission
+          <IconElement
+            variant={
+              selection === 'permission' && direction !== 'none'
+                ? `sort-${direction}`
+                : 'sort-indeterminate'
+            }
+          />
+        </ButtonElement>
+      ),
+      key: 'th-permission',
+      className: `${TABLE_HEADER_CLASS_NAME} ${TABLE_HEADER_CLASS_NAME}--type`,
+      onClick: () => onTableHeaderClick('permission'),
+      'aria-sort': selection === 'permission' ? direction : 'none',
+    },
+  ];
+
+  const compareFn = LocationsViewColumnSortMap[selection];
+
+  if (direction === 'ascending') {
+    data.sort((a, b) => compareFn(a[selection], b[selection]));
+  } else {
+    data.sort((a, b) => compareFn(b[selection], a[selection]));
+  }
+
+  const rows = data.map((location, index) => [
+    {
+      key: `td-scope-${index}`,
+      children: (
+        <ButtonElement
+          className={TABLE_DATA_BUTTON_CLASS}
+          onClick={() => onLocationClick(location)}
+          variant="table-data"
+        >
+          {location.scope}
+        </ButtonElement>
+      ),
+    },
+    { key: `td-type-${index}`, children: location.type },
+    { key: `td-permission-${index}`, children: location.permission },
+  ]);
+
+  return { columns, rows };
+};
+
+export function LocationsViewTableControl(): React.JSX.Element | null {
+  const [{ data, isLoading }] = useLocationsData();
+
+  const [, handleUpdateState] = useControl({ type: 'NAVIGATE' });
+
+  const [sortState, setSortState] = React.useState<
+    SortState<LocationAccess<Permission>>
+  >({
+    selection: 'scope',
+    direction: 'ascending',
+  });
+
+  const locationsData = React.useMemo(
+    () =>
+      getLocationsData({
+        data: data.result,
+        sortState,
+        onLocationClick: (payload) => {
+          handleUpdateState({
+            type: 'ACCESS_LOCATION',
+            location: payload,
+          });
+        },
+        onTableHeaderClick: (payload: keyof LocationAccess<Permission>) => {
```

Comment:
Think `payload` can just be `string`?

Reply:
I think setting it as `string` causes some type issues when I try to use `selection` to grab the compare function `const compareFn = LocationsViewColumnSortMap[selection]`. It says:


Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ scope: (a: string, b: string) => number; type: (a: string, b: string) => number; permission: (a: string, b: string) => number; }'.

Reply:
This is what I have at the moment:


```tsx
const getCompareFn = (selection: string) => {
  switch (selection) {
    case 'scope':
    case 'type':
    case 'permission':
      return compareStrings;
  }
};

const compareFn = getCompareFn(selection);

if (compareFn) {
  // not sure how to get it without any TS errors without casting as keyof LocationAccess
  const castSelection = selection as keyof LocationAccess;

  if (direction === 'ascending') {
    data.sort((a, b) => compareFn(a[castSelection], b[castSelection]));
  } else {
    data.sort((a, b) => compareFn(b[castSelection], a[castSelection]));
  }
}


```